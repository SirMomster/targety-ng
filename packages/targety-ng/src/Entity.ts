export interface EntityConfig<T> {
    metaData?: T;
}

export class Entity {
    public constructor(protected config: EntityConfig<unknown>) {}

    public getMetaData<T>(): T | undefined {
        return this.config.metaData as T | undefined;
    }
}
