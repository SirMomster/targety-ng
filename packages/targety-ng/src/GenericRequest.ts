import { Metadata } from ".";

export enum RequestType {
    EVENT = "EVENT",
    REQUEST = "REQUEST",
    AUTH_REQUEST = "AUTH_REQUEST",
    UNKNOWN = "UNKNOWN",
}

export class GenericRequest {
    private _metadata: Metadata = new Metadata();

    public isType(type: RequestType): boolean {
        return this.getType() === type;
    }

    /**
     * Returns the metadata for the given request.
     */
    public get metadata(): Metadata {
        return this._metadata;
    }

    public getType(): RequestType {
        return RequestType.UNKNOWN;
    }
}
