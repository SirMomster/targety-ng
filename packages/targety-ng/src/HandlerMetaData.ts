import { HttpMethod } from "./common/types";

// eslint-disable-next-line @typescript-eslint/no-empty-interface
export interface HandlerMetaDataConfig {}

export interface ExtenderHandlerMetaDataConfig<T> extends HandlerMetaDataConfig {
    metaData?: T;
}

export class HandlerMetaData {
    public constructor(private config: ExtenderHandlerMetaDataConfig<unknown>) {}

    public getMetaData<T>(): T | undefined {
        return this.config.metaData as T | undefined;
    }
}
