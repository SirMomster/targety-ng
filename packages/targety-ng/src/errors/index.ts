export * from "targety-ng-errors";
