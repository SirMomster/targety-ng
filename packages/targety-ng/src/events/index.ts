export * from "./EventStrategy";
export * from "./S3EventStrategy";
export * from "./SQSEventStrategy";
export * from "./AbstractEventStrategy";
