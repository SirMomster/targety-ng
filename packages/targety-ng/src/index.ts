import "reflect-metadata";

export { LambdaEntryPoint } from "./EntryPoint";
export { Handler } from "./Handler";
export { Middleware } from "./MiddlewareHandler";
export { Metadata } from "./Metadata";
export { GenericRequest } from "./GenericRequest";

export * as Util from "./utils";
export * as Error from "./errors";
export * as Enums from "./common/enums";
export * as Types from "./common/types";
export * as Interfaces from "./common/interfaces";

export * from "./OptionsHandler";

export * from "./common/decorators";

export * from "./handlers";
