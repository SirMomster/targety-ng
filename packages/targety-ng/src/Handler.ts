import { CLASS_METADATA } from "./common/Constants";
import { Request } from "./handlers/request/Request";
import { ResponseBody } from "./handlers/request/ResponseBody";
import { Middleware } from "./MiddlewareHandler";
import { Context } from "./common/interfaces";
import { HandlerMetaData } from "./HandlerMetaData";
import { RouteHandler } from "./handlers/request/RouteHandler";
import { EventHandler } from "./handlers/event/EventHandler";
import { GenericRequest, RequestType } from "./GenericRequest";
import { HandlerStrategy } from "./HandlerStrategy";
import { LambdaAuthorizerHandler } from "./handlers/auth/LambdaAuthorizerHandler";

interface HandlerInterface {
    [key: string]: (request: Request) => Promise<ResponseBody | void | unknown>;
}

export abstract class Handler implements HandlerInterface {
    [key: string]: any;
    /**
     * Middleware currently only applies to RequestType.REQUEST
     * Middleware is executed before the handler code. There are two types of middleware:
     *
     * 1) middleware that modifies the request object (express.js like functionality)
     * 2) middleware that returns a [[ResponseBody]] and interrupts the flow
     * 3) middleware that returns an object containing an onError and onSuccess
     *    function which can be resolved after the execution of the handler code.
     *
     * @protected
     * @type {Middleware[]}
     * @memberof Handler
     */
    public middleware: Middleware[] = [];
    public handlerMetaData: HandlerMetaData = this.getClassMetaData();

    private StrategyMap: Partial<Record<RequestType, HandlerStrategy<unknown, unknown>>> = {
        [RequestType.REQUEST]: new RouteHandler(this),
        [RequestType.EVENT]: new EventHandler(this),
        [RequestType.AUTH_REQUEST]: new LambdaAuthorizerHandler(this),
    };

    /**
     * Parses the action from the request, executes the middleware
     * and invokes the correct action on the handler implementation.
     */
    public async handle(request: GenericRequest, context?: Context): Promise<unknown> {
        const strategy = this.StrategyMap[request.getType()];

        if (!strategy) {
            throw new Error("No strategy found for request type " + request.getType());
        }

        return await strategy.handle(request, context);
    }

    protected getClassMetaData(): HandlerMetaData {
        const classMetaData = Reflect.getMetadata(CLASS_METADATA, this.constructor);
        return new HandlerMetaData({ metaData: classMetaData });
    }
}
