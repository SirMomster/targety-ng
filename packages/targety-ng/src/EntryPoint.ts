import { LOGGER } from "targety-ng-logger";
import { Objects } from "./utils";
import { Context } from "./common/interfaces";
import { Handler } from "./Handler";
import { Request } from "./handlers/request/Request";
import { LambdaProxyEvent } from "./handlers/request/LambdaProxyEvent";
import { Response } from "./handlers/request/Response";
import { ResponseBody } from "./handlers/request/ResponseBody";
import { GenericEvent } from "./handlers/event/GenericEvent";
import { GenericRequest, RequestType } from "./GenericRequest";
import { EventRequest } from "./handlers/event/EventRequest";
import { APIGatewayRequestAuthorizerEvent } from "aws-lambda";
import { LambdaAuthorizerRequest } from "./handlers/auth/LambdaAuthorizerRequest";

/**
 * Represents the LambdaEntryPoint abstract class
 */

export abstract class LambdaEntryPoint {
    public handler: Handler;
    public initPromise: Promise<Handler>;

    public constructor() {
        // pre-loading the handler, even when no request was made yet
        this.initPromise = this.initializeHandler();
    }

    /**
     * HeartBeatResponse can be used to keep the Lambda warm using Cloudwatch Events
     */
    private static heartBeatResponse(request: Request): ResponseBody {
        LOGGER.trace("Heartbeat function");
        // Disable cors because
        return Response.noContent(request, { cors: false, defaultHeaders: false }).send();
    }

    /**
     * Entrypoint for the Api Gateway Proxy event.
     */
    public async handle(
        event: LambdaProxyEvent | GenericEvent<unknown> | APIGatewayRequestAuthorizerEvent | {},
        context?: Context,
    ): Promise<unknown> {
        let finalRequest: GenericRequest;
        LOGGER.trace({ event, context, type: "REQUEST" }, "incoming request");

        if (context) {
            context.callbackWaitsForEmptyEventLoop = false;
        }

        try {
            if ((event as GenericEvent<unknown>).Records) {
                finalRequest = new EventRequest<GenericEvent<unknown>>(event as GenericEvent<unknown>);
            } else if ((event as APIGatewayRequestAuthorizerEvent).type === "REQUEST") {
                finalRequest = new LambdaAuthorizerRequest(event as APIGatewayRequestAuthorizerEvent);
            } else {
                const request = new Request(event as LambdaProxyEvent);
                finalRequest = request;
                if (Objects.isEmpty(event)) {
                    return LambdaEntryPoint.heartBeatResponse(request);
                }
            }
            return await this.handleRequest(finalRequest, context);
        } catch (e) {
            if (finalRequest?.isType(RequestType.REQUEST)) {
                LOGGER.error(e, "Error handling request");
                return Response.fromError(finalRequest as Request, e);
            }
            throw e;
        }
    }

    /**
     * initializeHandler is used for example to open a DB connection or set up Dependency Injection
     * @returns handler Handler exposes api endpoint methods
     */
    protected abstract initializeHandler(): Promise<Handler>;

    private async handleRequest(request: GenericRequest, context: Context): Promise<unknown> {
        if (!this.handler) {
            this.handler = await this.initPromise;
        }
        return this.handler.handle(request, context);
    }
}
