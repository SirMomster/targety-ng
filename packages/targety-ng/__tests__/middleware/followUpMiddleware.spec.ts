/* eslint-disable @typescript-eslint/explicit-function-return-type */
/* eslint-disable @typescript-eslint/ban-ts-ignore */
import { ApiGateWayProxyEvent } from "@dev/test-helper";
import "reflect-metadata";
import { GenericRequest } from "../../src";
import { ApiError } from "../../src/errors";
import { MiddlewareFollowUp } from "../../src/MiddlewareHandler";
import {
    CORS,
    DefaultCORS,
    Get,
    Handler,
    LambdaProxyEvent,
    Middleware,
    Request,
    ResponseBody,
    Response,
    S3,
} from "../../src";
import { ErrorCode } from "../../lib/errors";

const getProfileStub = jest.fn().mockReturnValue({ test: "data" });
const getS3Stub = jest.fn().mockReturnValue({ test: "data" });

const errorMiddleware: Middleware = async (): Promise<MiddlewareFollowUp> => {
    console.log("Loaded middleware");
    return {
        onError: async (request: GenericRequest, e: ApiError | Error): Promise<ResponseBody> => {
            return Response.forbidden(request as Request).send({ hidden: "The error is now hidden" });
        },
        onSuccess: async (request: GenericRequest, response: ResponseBody): Promise<ResponseBody> => {
            return response;
        },
    };
};

@DefaultCORS({ ExposedHeaders: [], AllowCredentials: true, AllowHeaders: ["x-api-key"] })
class HandlerImplementation extends Handler {
    public middleware: Middleware[] = [errorMiddleware];

    @Get("/test-endpont/{id}")
    @CORS({ ExposedHeaders: ["x-test"] })
    public async getProfile(req: Request) {
        throw new ApiError("some", 400, ErrorCode.Forbidden);
    }
}

let handlerImplementation: any;
let event: LambdaProxyEvent;
let request: any;

describe("Handler", () => {
    beforeEach(() => {
        handlerImplementation = new HandlerImplementation();
        event = ApiGateWayProxyEvent.get() as unknown as LambdaProxyEvent;
        event.httpMethod = "GET";
        event.resource = "/test-endpont/{id}";
        request = new Request(event);
    });

    afterEach(() => {
        jest.resetAllMocks();
    });

    it("is an abstract class that needs extending", () => {
        expect(handlerImplementation).toBeInstanceOf(Handler);
    });

    describe("setValidDomains", () => {
        const originalValue = process.env.NO_ALLOWED_DOMAINS;
        beforeAll(() => {
            process.env.NO_ALLOWED_DOMAINS = "false";
        });

        afterAll(() => {
            process.env.NO_ALLOWED_DOMAINS = originalValue;
        });
    });

    describe("#handle", () => {
        it("invokes correct path", async () => {
            const result = await handlerImplementation.handle(request);
            expect(result).toMatchSnapshot();
        });
    });
});
