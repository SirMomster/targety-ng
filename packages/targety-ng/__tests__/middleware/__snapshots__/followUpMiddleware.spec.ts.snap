// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Handler #handle invokes correct path 1`] = `
ResponseBody {
  "body": "{\\"hidden\\":\\"The error is now hidden\\"}",
  "multiValueHeaders": Object {
    "access-control-allow-credentials": Array [
      "true",
    ],
    "access-control-allow-origin": Array [
      "http://localhost:3000",
    ],
    "cache-control": Array [
      "no-store",
    ],
    "content-security-policy": Array [
      "'self'",
    ],
    "content-type": Array [
      "application/json",
    ],
    "strict-transport-security": Array [
      "max-age=31536000",
    ],
    "x-aws-request-id": Array [
      "c6af9ac6-7b61-11e6-9a41-93e8deadbeef",
    ],
    "x-content-type-options": Array [
      "nosniff",
    ],
    "x-frame-options": Array [
      "DENY",
    ],
  },
  "statusCode": 403,
}
`;
